#1731 Read the Docs configuration file v2

ggnaegi opened the pull request:
## Fixes #1731 
- #1731 

## Proposed Changes
  - add `.readthedocs.yaml` file to the root folder.
  - update to the latest sphinx version, update `conf.py`

Blog article: [Migrate your project to .readthedocs.yaml configuration file v2 — Read the Docs Blog](https://blog.readthedocs.com/migrate-configuration-v2/)

I just tried myself
https://ocelot-test.readthedocs.io/en/latest/features/authentication.html

ggnaegi commented:
@raman-m Ok, done, you can merge

raman-m commented:
> @raman-m Ok, done, you can merge

Not yet! 😉

raman-m commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,32 @@
+
+# Build documentation in the "docs/" directory with Sphinx
+sphinx:
+   configuration: docs/conf.py
```
We have to review and update this file!

ggnaegi replied:
@raman-m Ok, it was indeed very outdated, I have now updated the conf.py to the latest version (7.2.6)

raman-m reviewed (changes requested):
Technically adding mentioned **.yaml** file will resolve the problem of building, but let's update the `docs/conf.py` file too. I find the file very outdated.

raman-m commented:
### First,
The copyright message looks strange: https://ocelot-test.readthedocs.io/en/latest/index.html
Copyright is located at the bottom of each page!
![image](../assets/10501504/75f3bc7a-a297-4ddf-876f-a2d00418b2a3)
It is 2023 year now and project belongs to ThreeMammals org.
The line: [docs/conf.py | Line 62](../blob/develop/docs/conf.py#L62)

### Second,
Version of docs is 1.0.0 !!! 🤣 
![image](../assets/10501504/c91ac16c-6585-4aa7-85ae-a7c9d085a8ac)
So, HTML doc title contains this info:
![image](https://github.com/ThreeMammals/Ocelot/assets/10501504/6cf1ae9e-26ce-4901-be48-197367d2d2e6)
But actually version should be 20.0 because of current release 20.0
So, it is hardcoded now in config:
```python
# The short X.Y version.
version = '1.0.0'
```
The line: [docs/conf.py | Line 70](../blob/develop/docs/conf.py#L70)

### Third,
Release info is also wrong, it is 1.0.0 now. But I didn't find how it is reflected to documentation files.
```python
# The full version, including alpha/beta/rc tags.
release = '1.0.0'
```
The line: [docs/conf.py | Line 72](../blob/develop/docs/conf.py#L72)


### I propose to fix these issues. Hardcode for this hotfix is OK. But in future we have to write dynamic calculation of copyright year, version and release variables.

raman-m commented:
### Important blog article
[Doctools without configuration file (conf.py / mkdocs.yml) are deprecated — Read the Docs Blog](https://blog.readthedocs.com/doctool-without-configuration-file/) on 12 July 2023

raman-m commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,32 @@
+   configuration: docs/conf.py
+
+# Optionally build your docs in additional formats such as PDF and ePub
+# formats:
+#    - pdf
+#    - epub
```
We can try to add these formats. Otherwise we will have HTML docs only, right?

ggnaegi replied:
yes

raman-m commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,32 @@
+# See https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html
+# python:
+#    install:
+#    - requirements: docs/requirements.txt
```
And it seems wee could add this file too!
More: https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html

ggnaegi replied:
Ok, done

raman-m reviewed (changes requested):
Make corrections please!

ggnaegi commented:
@raman-m So, you can check it here: https://ocelot-test.readthedocs.io/en/bug-fix_documentation_generation

I have added the requirements.txt file, made some improvements in the doc (fixing documentation compilation warnings), fixed the copyrights and versions.

raman-m commented:
> (fixing documentation compilation warnings)

What are you talking about?

raman-m commented:
> @raman-m So, you can check it here: https://ocelot-test.readthedocs.io/en/bug-fix_documentation_generation
> 
> I have added the requirements.txt file, made some improvements in the doc (fixing documentation compilation warnings), fixed the copyrights and versions.

Thank you!

ggnaegi commented:
> > (fixing documentation compilation warnings)
> 
> What are you talking about?

When building the documentation locally...
![image](https://github.com/ThreeMammals/Ocelot/assets/58469901/a71ad111-a7b4-40d2-916e-53c0a265f9dd)

That's what I fixed.
```powershell
D:\dev\repositories\sphinx_test\building/releaseprocess.rst:19: WARNING: Enumerated list ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/authentication.rst:173: WARNING: Title underline too short.

Allowed Scopes
^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/authentication.rst:173: WARNING: Title underline too short.

Allowed Scopes
^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:18: WARNING: Title underline too short.

Claims to Claims Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:45: WARNING: Title underline too short.

Claims to Query String Parameters Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/claimstransformation.rst:45: WARNING: Title underline too short.

Claims to Query String Parameters Transformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/kubernetes.rst:31: WARNING: Explicit markup ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/kubernetes.rst:57: WARNING: Explicit markup ends without a blank line; unexpected unindent.
D:\dev\repositories\sphinx_test\features/loadbalancer.rst:105: WARNING: Title underline too short.

Custom Load Balancers
^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/loadbalancer.rst:105: WARNING: Title underline too short.

Custom Load Balancers
^^^^^^^^^^^^^^^^^^^^
D:\dev\repositories\sphinx_test\features/tracing.rst:2: WARNING: Duplicate explicit target name: "here".
D:\dev\repositories\sphinx_test\features/websockets.rst:76: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

raman-m commented:
> @ggnaegi commented [on Oct 15](#issuecomment-1763481237)

My local build results:
![image](https://github.com/ThreeMammals/Ocelot/assets/10501504/c7fe575f-6497-40e8-b8b4-be8fac4fb9fa)
I have 3 warnings! 😄

ggnaegi commented:
> > @ggnaegi commented [on Oct 15](#issuecomment-1763481237)
> 
> My local build results:
> ![image](https://github.com/ThreeMammals/Ocelot/assets/10501504/c7fe575f-6497-40e8-b8b4-be8fac4fb9fa)
> I have 3 warnings! 😄 

@raman-m Yeah, rebuild it again and it's going to be fine, only the static folder 😁

My concern is now the following: what about main branch?

raman-m commented:
> My concern is now the following: what about main branch?

Tomorrow I will create a PR to main from this release branch. The PR must be approved by Raynald. And then, we will get new release. And will check actual Docs build here: https://ocelot.readthedocs.io/
I will keep you updated...
